File: cs/Problems/Day15.cs
namespace AOC2023.Problems;

public class Day15 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => SumOfHashValues(input);

    private static int SumOfHashValues(ReadOnlySpan<char> input)
    {
        int sumOfHashes = 0;

        int hashValue = 0;
        for (int i = 0; i < input.Length; i++)
        {
            if (input[i] == ',')
            {
                sumOfHashes += hashValue;
                hashValue = 0;
                continue;
            }

            hashValue += input[i];
            hashValue = hashValue * 17 % 256;
        }
        sumOfHashes += hashValue;

        return sumOfHashes;
    }
}

File: go/problems/day15.go
package problems

type Day15 struct{}

func (d Day15) Solve(input string) int {
	return sumOfHashValues(input)
}

func sumOfHashValues(input string) int {
	sum := 0

	value := 0
	for i := 0; i < len(input); i++ {
		if input[i] == ',' {
			sum += value
			value = 0
			continue
		}

		value += int(input[i])
		value = value * 17 % 256
	}
	sum += value

	return sum
}
